C++/Navio/Common/gpio.cpp

```diff
@@ -137,37 +129,36 @@ void Pin::toggle()
     write(!read());
 }
 
+int Pin::getPeriBase() const
+{
+    int rpi_version = getRaspberryPiVersion();
+
+    switch (rpi_version) {
+    case 2:
+    case 3:
+        return BCM2708_PERI_BASE;
+    case 4:
+        return BCM2711_PERI_BASE;
+    default:
+        warnx("Unexpected version. Set PERI_BASE to default as for RPI2/3\n");
+        return BCM2708_PERI_BASE;
+    }
+}
+
 int Pin::getRaspberryPiVersion() const
 {
     char buffer[MAX_SIZE_LINE];
-    const char* hardware_description_entry = "Hardware";
-    const char* v1 = "BCM2708";
-    const char* v2 = "BCM2709";
-    const char* v3 = "BCM2835";
-    char* flag;
-    FILE* fd;
-
-    fd = fopen("/proc/cpuinfo", "r");
-
-    while (fgets(buffer, MAX_SIZE_LINE, fd) != NULL) {
-        flag = strstr(buffer, hardware_description_entry);
-
-        if (flag != NULL) {
-            if (strstr(buffer, v3) != NULL) {
-                fclose(fd);
-                return 3;
-            } else if (strstr(buffer, v2) != NULL) {
-                fclose(fd);
-                return 2;
-            } else if (strstr(buffer, v1) != NULL) {
-                fclose(fd);
-                return 1;
-            }
+    int rpi_version;
+
+    FILE *f = fopen("/sys/firmware/devicetree/base/model", "r");
+    if (f != 0 && fgets(buffer, MAX_SIZE_LINE, f) != 0) {
+        int ret = sscanf(buffer + 12, "%d", &rpi_version);
```

Comment:
Magic number. Are we sure that this offset will always be the same?

Reply:
It should be because it's common for all RPi's versions, 
`cat /sys/firmware/devicetree/base/model` gives you something like this:
`Raspberry Pi 4 Model B Plus Rev 1.3`

It is changed only version, model and revision, so the offset would always be there.

Reply:
Magic number. Add define then